### DISCUSSION TITLE
Clarification on Fee Tracking in Positions Management

### DISCUSSION THREAD

**Discussion by allenzhao** [Jeiwan/uniswapv3-book#36] [Milestone 5. Fees and Price Oracle] [answered]
Hi, first I would like to thank you for providing a detailed explanation of the Uniswap V3 code!

I was wondering if you could help clarify a few points, I'm not sure if I understand them correctly (I'm asking these because I'm trying to calculate positions' fees accumulated during a time period, using Uniswap's hosted subgraph and the time-travel queries based on blocks):

1. Are `feeGrowthGlobal1X128` and `feeGrowthGlobal1X128` always non-decreasing for a given pool?  I cannot find a place that these were changed by directly searching, am I missing something? I understand that every time when a trade happens, there will be fees added, but is there any possibility that  `feeGrowthGlobal1X128` or `feeGrowthGlobal1X128` will be decreased?
2. At any moment in time, if I know all the required state variables used in `getFeeGrowthInside`, is it possible to calculate how much fees a position would have received per liquidity till this time by directly using `feeGrowthInside0X128` * `position.liquidity`?  This is different from the `tokensOwed0` and `tokensOwed1` calculation, correct?
3. The `getFeeGrowthInside` function is called when the LP modifies the position, based on the `_modifyPosition` function in [here](https://github.com/Jeiwan/uniswapv3-code/tree/ed23a4ba344dafc8eacc417fce14212512b91bf7/src/UniswapV3Pool.sol). I was wondering, if without the `_modifyPosition`, and it seems that along with the position updates, there are also updates to the ticks [here](https://github.com/Jeiwan/uniswapv3-code/blob/ed23a4ba344dafc8eacc417fce14212512b91bf7/src/UniswapV3Pool.sol#L186-L201) that might change the `feeGrowthOutside1X128` or `feeGrowthOutside1X128`, when `liquidityBefore` is 0 and `tick` < `currentTick`. Is this related to the above question that I asked? Is it possible that some of the state variables used in  `getFeeGrowthInside` function being different based on some missing state updates?

Thanks in advance!

**Comment by Jeiwan** [comment, marked as answer]
> Are feeGrowthGlobal1X128 and feeGrowthGlobal1X128 always non-decreasing for a given pool?

Yes. This is so because swap fees only grow up: trades are made, swap fees are taken. These variables are updated during swapping:
https://github.com/Uniswap/v3-core/blob/main/contracts/UniswapV3Pool.sol#L760-L766
When swapping tokens, swap fees are subtracted from input amounts and counted. After a swap is done, the global fees are increased by the total swap fees accumulated during the swap.

> is it possible to calculate how much fees a position would have received per liquidity till this time by directly using feeGrowthInside0X128 * position.liquidity

If I understand the question correctly, it's not different from the owed tokens calculation. `feeGrowthInside0X128` * `position.liquidity` is used to find actual token amounts accrued as fees by a position. `tokensOwed0` and `tokensOwed1` are these amounts. But we can only calculate and save them when updating a position, to save gas. `getFeeGrowthInside` calculates fee growth *per liquidity*.

If you know how lending protocols work, there's a similarity: interest must always be accrued before modifying loans. Here we must always accrue swap fees (convert them to token amounts) before updating liquidity position. Also notice these lines:
https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Position.sol#L64
https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Position.sol#L72
Owed tokens are calculated on the changes of `feeGrowthInside0X128`/`feeGrowthInside1X128`.

I think this will answer your third question:
1. During swapping, collected swap fees are saved to the global fee tracking variables. They are saved as "fee per liquidity" (notice the division here https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/UniswapV3Pool.sol#L690).
1. When ticks are crossed (the current prices enters or leaves a position), fee growth outside a position is saves (https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Tick.sol#L178-L179).
1. Before a position is updated (liquidity is added or removed), fee growth inside the position is calculated (https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/UniswapV3Pool.sol#L439-L440) (which is calculated based on fee growth outside of the position, https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Tick.sol#L75-L90). "Fee growth inside a position" basically means "fees per liquidity accumulated by a position".
1. When a position is updated, *fees are accrued*, or, in other words, fee growth inside a position is turned into actual owed token amounts (https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Position.sol#L61-L85). Notice that these variables are updated: https://github.com/Uniswap/v3-core/blob/fc2107bd5709cdee6742d5164c1eb998566bcb75/contracts/libraries/Position.sol#L80-L81

> Is it possible that some of the state variables used in getFeeGrowthInside function being different based on some missing state updates?

I don't think so. It's a conveyor:
1. first, there are global fee trackers;
1. then there are ticks that track fees accumulated outside of them;
1. then there are positions (made of two ticks) that have fees accumulated inside of them (between the ticks);
1. then there are actual token amounts that are calculated when position's liquidity is changed (fees accruing).

Looks solid :)

(sorry for the typos, it's a late Friday night here :)
